scripts/split_repo_migration/git-migrate-history.sh

```diff
@@ -0,0 +1,156 @@
+  $EXIT 1
+fi
+
+# source GitHub repository. format: <owner>/<repo>
+SOURCE_REPO=$1
+
+# destination GitHub repository. format: <owner>/<repo>
+TARGET_REPO=$2
+
+# path in the source repo to copy code from. Defaults to the root directory
+SOURCE_PATH=$3
+
+# path in the target repo to put the copied code
+TARGET_PATH=$4
+
+# comma-separated list of files/folders to skip
+IGNORE_FOLDERS=$5
+# keep these specific files that would otherwise be deleted by IGNORE_FOLDERS
+KEEP_FILES=$6
+
+# override the HEAD branch name for the migration PR
+BRANCH=$7
+
+
+if [[ -z "${BRANCH}" ]]
+then
+  # default the branch name to be generated from the source repo name
+  BRANCH=$(basename ${SOURCE_REPO})-migration
+fi
+export FILTER_BRANCH_SQUELCH_WARNING=1
+
+# create a working directory
+WORKDIR="$(mktemp -d -t code-migration-XXXXXXXXXX)"
+echo "Created working directory: ${WORKDIR}"
+
+pushd "${WORKDIR}"  # cd into workdir
+
+echo "Cloning source repository: ${SOURCE_REPO}"
+git clone "git@github.com:${SOURCE_REPO}.git" source-repo
+
+pushd source-repo
+git remote remove origin
+
+# prune only files within the specified directory
+if [[ ! -z "${SOURCE_PATH}" ]]
+then
+  echo "Pruning commits only including path: ${SOURCE_PATH}"
+  git filter-branch \  # rewrites history... unsafe
+    --prune-empty \
+    --subdirectory-filter "${SOURCE_PATH}"
+fi
+
+if [[ ! -z "${IGNORE_FOLDERS}" ]]
+then
+  echo "Ignoring folder: ${IGNORE_FOLDERS}"
+  mkdir -p "${WORKDIR}/filtered-source"
+  FOLDERS=$(echo ${IGNORE_FOLDERS} | tr "," " ")
+  # remove files/folders we don't want
+  FILTER="(rm -rf ${FOLDERS} || true)"
+  if [[ ! -z "${KEEP_FILES}" ]]
+  then
+    KEEP_FILES_SPACES=($(echo ${KEEP_FILES} | tr "," " "))
+    LAST_ELEMENT=$(( ${#KEEP_FILES_SPACES[@]} - 1 ))
+    KEEP_FILE_COMMANDS=""
+    for file in "${KEEP_FILES_SPACES[@]}"
+    do
+      if [[ $file == "${KEEP_FILES_SPACES[$LAST_ELEMENT]}" ]]
+      then
+        KEEP_FILE_COMMANDS+="git checkout -- ${file} 2>/dev/null || true"
+      else 
+        KEEP_FILE_COMMANDS+="git checkout -- ${file} 2>/dev/null || true; "
+      fi   
+    done
+    # restore files to keep, silence errors if the file doesn't exist
+    FILTER="${FILTER}; ${KEEP_FILE_COMMANDS}"
+  fi
+  git filter-branch \
+    --force \
+    --prune-empty \
+    --tree-filter "${FILTER}"
+fi
+
+# reorganize the filtered code into the desired target locations
+if [[ ! -z "${TARGET_PATH}" ]]
+then
+  echo "Moving files to destination path: ${TARGET_PATH}"
+  git filter-branch \
+    --force \
+    --prune-empty \
+    --tree-filter \
+      "shopt -s dotglob; mkdir -p ${WORKDIR}/migrated-source; mv * ${WORKDIR}/migrated-source; mkdir -p ${TARGET_PATH}; mv ${WORKDIR}/migrated-source/* ${TARGET_PATH}"
+fi
+
+# back to workdir
+popd
+
+# merge histories
+echo "Cloning target repository: ${SOURCE_REPO}"
+git clone "git@github.com:${TARGET_REPO}.git" target-repo
```

Comment:
We should talk. I propose you don't clone $TARGET_REPO, but allow it to be provided as a parameter to run.sh, and here you just assume it exists. If you do this, then in my batch update script I can clone $TARGET_REPO once, and use `git worktree` to create a new workdir directory referencing that cloned repo with an API-specific branch created from `main`. Then the rest of this script can import the changes in that directory. 

This way we can easily both generate multiple per-API merge PRs simumltaneously AND not commit google-cloud-python repeatedly. WDYT? Happy to talk more.

Reply:
Sounds good. I no longer clone TARGET_REPO and instead accept path to target repo as an argument.